File: medium/139. Word Break/139. Word Break.ts
function wordBreak(s: string, wordDict: string[]): boolean {
    const wordSet = new Set(wordDict)
    const n = s.length
    const dp: boolean[] = Array(n + 1).fill(false)
    dp[0] = true

    for (let i = 1; i < n + 1; i++) {
        for (let j = 0; j < i; j++) {
            if (dp[j] && wordSet.has(s.slice(j, i))) {
                dp[i] = true
                break
            }
        }
    }

    return dp[n]
};


console.log(wordBreak("leetcode", ["leet", "code"]))

File: medium/139. Word Break/139. Word Break.py
from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word_set = set(wordDict)
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(1, n + 1):
            for j in range(i):
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break

        return dp[n]


solution = Solution()

print(solution.wordBreak(s="leetcode", wordDict=["leet", "code"]))
print(solution.wordBreak(s="applepenapple", wordDict=["apple", "pen"]))
print(solution.wordBreak(s="catsandog", wordDict=["cats", "dog", "sand", "and", "cat"]))
print(solution.wordBreak(s="aaaaaaa", wordDict=["aaaa", "aaa"]))
print(solution.wordBreak(s="a", wordDict=["b"]))
